posts/tradeoff-of-defining-types-as-subobjects.md

```diff
@@ -0,0 +1,119 @@
+```lean
+-- Subobject design
+def foo : circle where
+  val := 1
+  property := by simp
+
+-- Coerced subobject design
+def bar : Circle where
+  val := 1
+  property := by simp
+```
+
+In the Custom Structure design, projections can be custom-named, leading to more informative code:
+```lean
+def baz : Circle where
+  val := 1
+  norm_val := by simp
+```
+
+## Generic instances
+
+In the Coerced Subobject and Custom Structure designs, generic instances about subobjects do not apply and must be copied over manually. In the Coerced Subobject design, they can easily be transferred/derived:
+```lean
+def Circle : Type := circle
+deriving TopologicalSpace
+
+instance : MetricSpace Circle := Subtype.metricSpace
+```
+
+In the Custom Structure design, these instances must be either copied over manually or transferred some kind of isomorphism.
+
+The Subobject design, by definition, lets all generic instances apply.
+
+## Conclusion
+
+Here is a table compiling the above discussion.
+
+| Aspect | Subobject | Coerced Subobject | Custom Structure |
+|--|--|--|--|
+| Typeclass search | Possibly expensive | Fast | Fast |
+| Dot notation | ✗ | ✓ | ✓ |
+| Custom named projections | ✗ | ✗ | ✓ |
+| Generic instances | automatic | easy to transfer | hard to transfer but could be improved |
+
+In conclusion, the Custom Structure design is superior in performance and usability at the expense of a one-time overhead cost when transferring generic instances. The Subobject design is still viable for types that see little use, so as to avoid the instance transferring cost. Finally, the Coerced Subobject is a good middle ground when having custom named projections is not particularly necessary. It could also reasonably be used as an intermediate step when refactoring from the Subobject design to the Custom Structure design.
+
+## Further concerns
+
+Subobjects are not the only ones having a coercion to `Type`: Concrete categories are categories whose objects can be interpreted as types, and there usually is a coercion from such a category to `Type`.
+
+For example, [`AlgebraicGeometry.Scheme`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=AlgebraicGeometry.Scheme#doc) sees widespread use in algebraic geometry and [`AlgebraicGeometry.Proj`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=AlgebraicGeometry.Proj#doc) is a term of type `Scheme` that is also used as a type.
+
+This raises issues of its own, as `Proj` is (mathematically) also a smooth manifold. Since terms only have one type, we can't hope to have both `Proj : Scheme` and `Proj : SmoothMnfld` (this last category does not exist yet in Mathlib). One option would be to have a separate `DifferentialGeometry.Proj` of type `SmoothMnfld`. Another one would be to provide the instances saying that `AlgebraicGeometry.Proj` is a smooth manifold.
```

Comment:
Only the `Proj` of *some* rings are smooth manifolds.  A random (graded) ring will typically have singularities.